docs/notebooks/flax_basics.ipynb

```diff
@@ -0,0 +1,911 @@
+{
```

Comment:
<p>First point is not true. Even in your example notice that you created <code>Dense(3)</code>  instance with only the output shape. With <code>setup()</code>  you can either have shape inference or not.</p><p> </p><p>An important point is that if you want more than one method on a module, then you <em>must</em> use <code>setup</code>  (we disallow two methods using <code>nn.compact</code>)  -- this is to avoid many hairy issues about whether a module is reused or not, or how they are named if you call one method from another</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/542/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Reply:
<p>Ah, I'm confused then because that was my understanding from the linen notebook (the alpha one) with this sentence:</p><blockquote>We can also declare variables in setup, though in doing so you can't take advantage of shape inference and have to provide explicit shape information at initialization. The syntax is a little repetitive in this case right now, but we do force agreement of the assigned names.</blockquote><p>Let's chat to make sure I understand correctly. For the second point on multiple method, added a note.</p>

Reply:
<p>Add a reference to a doc needed for patterns with shape inference (or not)</p>